File: ports/jvm/src/main/java/com/unicodesecurity/IdentifierTokens.java
package com.unicodesecurity;

import java.util.ArrayList;
import java.util.List;

/**
 * Identifier-shaped tokens of a running-text field.
 *
 * <p>Direct port of {@code Unicode/Security/Identity/IdentifierTokens.lean}. A
 * source line, a chat message or a display name is not one identifier, but the
 * identifier-shaped words inside it are the surface a homoglyph attack targets
 * ({@code scоpe} with a Cyrillic о in {@code let scоpe = 1;}). A token is a
 * maximal run of {@code XID_Continue} codepoints; everything else (space,
 * punctuation, operators, format controls) separates tokens. Each token carries
 * its start position so a finding made on the token can be reported in input
 * coordinates.
 *
 * <p>{@code XID_Continue} is the port's own {@link Security#isXidContinue} over
 * the bundled DerivedCoreProperties.txt.
 */
public final class IdentifierTokens {
  private IdentifierTokens() {}

  /** One identifier-shaped run: its start position in the input and its codepoints. */
  public record Token(int start, List<Integer> cps) {}

  /** The maximal {@code XID_Continue} runs of the input, in input order. */
  public static List<Token> tokens(List<Integer> input) {
    List<Token> out = new ArrayList<>();
    int start = 0;
    List<Integer> current = null;
    for (int idx = 0; idx < input.size(); idx++) {
      int cp = input.get(idx);
      if (Security.isXidContinue(cp)) {
        if (current == null) {
          current = new ArrayList<>();
          start = idx;
        }
        current.add(cp);
      } else if (current != null) {
        out.add(new Token(start, List.copyOf(current)));
        current = null;
      }
    }
    if (current != null) {
      out.add(new Token(start, List.copyOf(current)));
    }
    return List.copyOf(out);
  }

  /** Move token-local positions back into input coordinates. */
  public static List<Integer> shiftPositions(int start, List<Integer> positions) {
    List<Integer> out = new ArrayList<>();
    for (int pos : positions) {
      out.add(pos + start);
    }
    return List.copyOf(out);
  }
}

File: ports/dotnet/src/UnicodeSecurity/Identity/IdentifierTokens.cs
namespace UnicodeSecurity;

// identifier-tokens — identifier-shaped tokens of a running-text field.
//
// Direct port of Unicode/Security/Identity/IdentifierTokens.lean. A source
// line, a chat message or a display name is not one identifier, but the
// identifier-shaped words inside it are the surface a homoglyph attack targets
// (scоpe with a Cyrillic о in `let scоpe = 1;`). A token is a maximal run of
// XID_Continue codepoints; everything else (space, punctuation, operators,
// format controls) separates tokens. Each token carries its start position so a
// finding made on the token can be reported in input coordinates.
//
// XID_Continue is the port's own IsXidContinue over the bundled
// DerivedCoreProperties.txt.
public static partial class Security
{
    public static class IdentifierTokens
    {
        /// <summary>One identifier-shaped run: its start position in the input
        /// and its codepoints.</summary>
        public sealed record Token(int Start, List<int> Cps);

        /// <summary>The maximal XID_Continue runs of the input, in input
        /// order.</summary>
        public static List<Token> Tokens(IReadOnlyList<int> input)
        {
            var tokens = new List<Token>();
            var start = 0;
            List<int>? current = null;
            for (var idx = 0; idx < input.Count; idx++)
            {
                var cp = input[idx];
                if (Security.IsXidContinue(cp))
                {
                    if (current is null)
                    {
                        current = new List<int>();
                        start = idx;
                    }
                    current.Add(cp);
                }
                else if (current is not null)
                {
                    tokens.Add(new Token(start, current));
                    current = null;
                }
            }
            if (current is not null)
            {
                tokens.Add(new Token(start, current));
            }
            return tokens;
        }

        /// <summary>Move token-local positions back into input
        /// coordinates.</summary>
        public static List<int> ShiftPositions(int start, IReadOnlyList<int> positions) =>
            positions.Select(p => p + start).ToList();
    }
}
